### PR TITLE
server: unix socket should verify user's authentication

### LINKED ISSUE #8372: tidb-server does not correctly handle unix-socket connections
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
```
# session1
./bin/tidb-server --socket=/tmp/tidbtmp.sock
```
```
# session2
mysql -S /tmp/tidbtmp.sock -e 'select user()'
ERROR 1105 (HY000) at line 1: Missing session variable when eval builtin
```

In addition, the only schema visible is `INFORMATION_SCHEMA`.  Running a command like `show grants` segfaults because the session is not configured.

2. What did you expect to see?

Show current user.

3. What did you see instead?

Error log:
```
morgo@ryzen:~/Downloads/tidb-latest-linux-amd64$ ./bin/tidb-server --socket=/tmp/tidbtmp.sock
2018/11/20 06:29:07.361 printer.go:39: [info] Welcome to TiDB.
2018/11/20 06:29:07.361 printer.go:40: [info] Release Version: v2.1.0-rc.3-199-gc84537d
2018/11/20 06:29:07.361 printer.go:41: [info] Git Commit Hash: c84537dc944853c02c2ead9500cedf681eb4a175
2018/11/20 06:29:07.361 printer.go:42: [info] Git Branch: master
2018/11/20 06:29:07.361 printer.go:43: [info] UTC Build Time:  2018-11-19 06:24:07
2018/11/20 06:29:07.361 printer.go:44: [info] GoVersion:  go version go1.11 linux/amd64
2018/11/20 06:29:07.361 printer.go:45: [info] Race Enabled: false
2018/11/20 06:29:07.361 printer.go:46: [info] Check Table Before Drop: false
2018/11/20 06:29:07.361 printer.go:47: [info] TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
2018/11/20 06:29:07.362 printer.go:52: [info] Config: {"host":"0.0.0.0","advertise-address":"","port":4000,"cors":"","store":"mocktikv","path":"/tmp/tidb","socket":"/tmp/tidbtmp.sock","lease":"45s","run-ddl":true,"split-table":true,"token-limit":1000,"oom-action":"log","mem-quota-query":34359738368,"enable-streaming":false,"txn-local-latches":{"enabled":true,"capacity":2048000},"lower-case-table-names":2,"log":{"level":"info","format":"text","disable-timestamp":false,"file":{"filename":"","log-rotate":true,"max-size":300,"max-days":0,"max-backups":0},"slow-query-file":"","slow-threshold":300,"expensive-threshold":10000,"query-log-max-len":2048},"security":{"skip-grant-table":false,"ssl-ca":"","ssl-cert":"","ssl-key":"","cluster-ssl-ca":"","cluster-ssl-cert":"","cluster-ssl-key":""},"status":{"report-status":true,"status-port":10080,"metrics-addr":"","metrics-interval":15},"performance":{"max-procs":0,"tcp-keep-alive":true,"cross-join":true,"stats-lease":"3s","run-auto-analyze":true,"stmt-count-limit":5000,"feedback-probability":0.05,"query-feedback-limit":1024,"pseudo-estimate-ratio":0.8,"force-priority":"NO_PRIORITY"},"xprotocol":{"xserver":false,"xhost":"","xport":0,"xsocket":""},"prepared-plan-cache":{"enabled":false,"capacity":100},"opentracing":{"enable":false,"sampler":{"type":"const","param":1,"sampling-server-url":"","max-operations":0,"sampling-refresh-interval":0},"reporter":{"queue-size":0,"buffer-flush-interval":0,"log-spans":false,"local-agent-host-port":""},"rpc-metrics":false},"proxy-protocol":{"networks":"","header-timeout":5},"tikv-client":{"grpc-connection-count":16,"grpc-keepalive-time":10,"grpc-keepalive-timeout":3,"commit-timeout":"41s","max-txn-time-use":590},"binlog":{"enable":false,"write-timeout":"15s","ignore-error":false},"compatible-kill-query":false}
2018/11/20 06:29:07.362 main.go:198: [info] disable Prometheus push client
2018/11/20 06:29:07.362 systime_mon.go:24: [info] start system time monitor
2018/11/20 06:29:07.362 tidb.go:272: [info] new store
2018/11/20 06:29:07.384 tidb.go:62: [info] store 4143ee46-c114-418d-8c06-4f84fbf001c5 new domain, ddl lease 45s, stats lease 3000000000
2018/11/20 06:29:07.385 ddl.go:362: [info] [ddl] start DDL:233cf12b-d4d7-4d1f-82d4-24c3a32f6c26, run worker true
2018/11/20 06:29:07.385 ddl_worker.go:82: [info] [ddl] start delRangeManager OK, with emulator: true
2018/11/20 06:29:07.386 ddl_worker.go:82: [info] [ddl] start delRangeManager OK, with emulator: true
2018/11/20 06:29:07.386 ddl_worker.go:117: [info] [ddl-worker 2, tp add index] start DDL worker
2018/11/20 06:29:07.386 ddl_worker.go:117: [info] [ddl-worker 1, tp general] start DDL worker
2018/11/20 06:29:07.387 delete_range.go:114: [info] [ddl] start delRange emulator
2018/11/20 06:29:07.387 delete_range.go:114: [info] [ddl] start delRange emulator
2018/11/20 06:29:07.394 domain.go:125: [info] [ddl] full load InfoSchema from version 0 to 15, in 7.852517ms
2018/11/20 06:29:07.394 domain.go:322: [info] [ddl] full load and reset schema validator.
2018/11/20 06:29:07.402 server.go:194: [warning] Secure connection is NOT ENABLED
2018/11/20 06:29:07.402 server.go:156: [info] Server is running MySQL Protocol through Socket [/tmp/tidbtmp.sock]
2018/11/20 06:29:07.403 domain.go:757: [info] [stats] init stats info takes 1.293653ms
2018/11/20 06:29:07.403 http_status.go:119: [info] Listening on :10080 for status and metrics report.
2018/11/20 06:29:17.850 server.go:314: [info] con:1 new connection @
2018/11/20 06:29:48.783 constant_fold.go:136: [warning] fold constant user(): Missing session variable when eval builtin
2018/11/20 06:29:48.783 conn.go:488: [warning] con:1 dispatch error:
id:1, addr:@ status:2, collation:utf8_general_ci, user:morgo
"select user()"
Missing session variable when eval builtin
github.com/pingcap/tidb/expression.(*builtinUserSig).evalString
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_info.go:188
github.com/pingcap/tidb/expression.(*ScalarFunction).EvalString
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:230
github.com/pingcap/tidb/expression.executeToString
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:213
github.com/pingcap/tidb/expression.evalOneColumn
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/chunk_executor.go:96
github.com/pingcap/tidb/expression.(*defaultEvaluator).run
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/evaluator.go:48
github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/expression/evaluator.go:114
github.com/pingcap/tidb/executor.(*ProjectionExec).unParallelExecute
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/projection.go:167
github.com/pingcap/tidb/executor.(*ProjectionExec).Next
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/projection.go:152
github.com/pingcap/tidb/executor.(*recordSet).Next
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go:97
github.com/pingcap/tidb/server.(*tidbResultSet).Next
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/driver_tidb.go:359
github.com/pingcap/tidb/server.(*clientConn).writeChunks
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1000
github.com/pingcap/tidb/server.(*clientConn).writeResultset
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:964
github.com/pingcap/tidb/server.(*clientConn).handleQuery
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:881
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:626
github.com/pingcap/tidb/server.(*clientConn).Run
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:470
github.com/pingcap/tidb/server.(*Server).onConn
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:324
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1333

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v2.1.0-rc.3-199-gc84537d
Git Commit Hash: c84537dc944853c02c2ead9500cedf681eb4a175
Git Branch: master
UTC Build Time: 2018-11-19 06:24:07
GoVersion: go version go1.11 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Fix #8372 . The connection should verify user's authentication when using unix socket.

### What is changed and how it works?
Verify user's authentication when using unix socket.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8381)

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -129,7 +129,7 @@ func (s *Server) newConn(conn net.Conn) *clientConn {
 	return cc
 }
 
-func (s *Server) skipAuth() bool {
+func (s *Server) isLocal() bool {
```
Comment (tiancaiamao) [inline server/server.go]: The function name is `isLocal` and the function implement is `c.cfg.Socket` ? That seems weird 
@jackysp

Comment (morgo) [reply server/server.go]: Should it be `isLocalSocket` or `isUnixSocket`?  There could be a local `tcp` connection.  We will not support them, but MySQL also supports local named pipe and shm on windows.

Comment (jackysp) [reply server/server.go]: `isUnixSocket` sounds good.

File: server/conn.go [hunk 2]
```diff
@@ -391,16 +392,17 @@ func (cc *clientConn) openSessionAndDoAuth(authData []byte) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if !cc.server.skipAuth() {
-		// Do Auth.
+	host := variable.ServerHostname
```
Comment (lysu) [inline server/conn.go]: it seems mysql here will got "localhost" and do auth logic

Comment (jackysp) [reply server/conn.go]: Yes, I know it. I used to set host to a const string here, due to I have not checked the implementation of MySQL. If it is critical, I'll check when will MySQL set "localhost" and whether there are other cases.

Comment (morgo) [reply server/conn.go]: I think a const is valid here :-) MySQL has a special handling/meaning of "localhost".  Clients that connect to "localhost" will be redirected to a socket etc.

**Comment by jackysp** [comment]
PTAL @tiancaiamao @lysu @zimulala

**Pull request merged**